### PR TITLE
feat: Implement custom repo filtering and sort by modified date

### PR DESCRIPTION
**Pull request by arran4**
Adds columns for "Created", "Archived", and "Fork" to the repository list view, implements dynamic filtering based on a query language, defaults to hiding forks and archived repos, and sorts the list by modified date by default.

---
*PR created automatically by Jules for task [18354772602961063441](https://jules.google.com/task/18354772602961063441) started by @arran4*

### CODE REVIEW

File: src/utils/FilterParser.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    for (const auto& child : m_children) parts.append(child->toString());
+    return QStringLiteral("(") + parts.join(QStringLiteral(" AND ")) + QStringLiteral(")");
+}
+
+bool OrNode::evaluate(const FilterDataAccessor& accessor) const {
+    for (const auto& child : m_children) {
+        if (child->evaluate(accessor)) return true;
+    }
+    return m_children.isEmpty();
+}
```
[the remark this answers is not in the document]
Comment (arran4) [reply src/utils/FilterParser.cpp:202]: > medium
> 
> In boolean logic, the identity element for an OR operation is `false`. An empty `OrNode` should evaluate to `false`, whereas an empty AND (correctly implemented at line 161) evaluates to `true`. While the current parser implementation might not produce empty `OrNode` objects, this implementation is logically incorrect and could lead to unexpected behavior if the parser is modified in the future.
> Suggested change
> bool OrNode::evaluate(const FilterDataAccessor& accessor) const {
>     for (const auto& child : m_children) {
>         if (child->evaluate(accessor)) return true;
>     }
>     return m_children.isEmpty();
> }
> bool OrNode::evaluate(const FilterDataAccessor& accessor) const {
>     for (const auto& child : m_children) {
>         if (child->evaluate(accessor)) return true;
>     }
>     return false;
> }

@jules

File: src/utils/FilterParser.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    if (lowerKey == QStringLiteral("repo") || lowerKey == QStringLiteral("owner")) {
+#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
+        QRegularExpression re = QRegularExpression::fromWildcard(m_value, Qt::CaseInsensitive);
+        return re.match(val).hasMatch();
+#else
+        QRegExp rx(m_value);
+        rx.setPatternSyntax(QRegExp::Wildcard);
+        rx.setCaseSensitivity(Qt::CaseInsensitive);
+        return rx.exactMatch(val);
+#endif
```
[the remark this answers is not in the document]
Comment (arran4) [reply src/utils/FilterParser.cpp]: > high
> 
> The use of `QRegularExpression::fromWildcard` was introduced in Qt 6.3. Since the `CMakeLists.txt` only requires `Qt6` without a minor version, this code will fail to compile on older Qt 6 versions (e.g., 6.0 through 6.2). Additionally, `QRegExp` is deprecated in Qt 5 and removed in Qt 6. It is recommended to use `QRegularExpression::wildcardToRegularExpression` for compatibility with all Qt 5 and Qt 6 versions, which is also consistent with the implementation used in `NotificationRuleEngine.cpp`.
> Suggested change
> #if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
>         QRegularExpression re = QRegularExpression::fromWildcard(m_value, Qt::CaseInsensitive);
>         return re.match(val).hasMatch();
> #else
>         QRegExp rx(m_value);
>         rx.setPatternSyntax(QRegExp::Wildcard);
>         rx.setCaseSensitivity(Qt::CaseInsensitive);
>         return rx.exactMatch(val);
> #endif
> #if QT_VERSION >= QT_VERSION_CHECK(6, 3, 0)
>         QRegularExpression re = QRegularExpression::fromWildcard(m_value, Qt::CaseInsensitive);
>         return re.match(val).hasMatch();
> #else
>         QRegularExpression re(QRegularExpression::wildcardToRegularExpression(m_value), QRegularExpression::CaseInsensitiveOption);
>         return re.match(val).hasMatch();
> #endif

@jules

**Comment by arran4** [comment]
@jules I see the parser but I don't see how it's integrated or used.

**Comment by arran4** [comment]
@jules the parser seems to have disappeared can you pelaase re add it and add test cases

**Comment by arran4** [comment]
@jules
```
src/utils/FilterParser.cpp:217:2: error: failed to evaluate #if condition, undefined function-like macro invocation: QT_VERSION_CHECK( ... ) [syntaxError]
#if QT_VERSION >= QT_VERSION_CHECK(6, 3, 0)
 ^
```

File: tests/TestFilterParser.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+class TestFilterParser : public QObject {
+    Q_OBJECT
+   private slots:
+    void testBasicFilter() {
+        QJsonObject repo;
+        repo["fork"] = false;
+        repo["archived"] = false;
+        repo["name"] = "test-repo";
+
+        QSharedPointer<ASTNode> ast = FilterParser::parse("fork:false AND archived:false");
```
Comment (arran4) [inline tests/TestFilterParser.cpp:38]: @jules add a to string version so we can test round parsing, also please add bracket support, please make the tests extensive too with edge cases also self simplifying (removing brackets where they are not needed, doubled, presidence etc.)

Comment (arran4) [reply tests/TestFilterParser.cpp:38]: @jules please greatly extend the number of tests.

File: tests/TestFilterParser.cpp
Comment (arran4) [inline tests/TestFilterParser.cpp:1]: @jules please expand the number of test cases, looking for edge cases like NOT NOT, nested brackets, etc.

**Comment by arran4** [comment]
@jules resolve:
```
Run find src tests \( -name '*.cpp' -o -name '*.cc' -o -name '*.h' -o -name '*.hpp' \) -print0 | xargs -0 -r clang-format --dry-run --Werror
tests/TestFilterParser.cpp:137:115: error: code should be clang-formatted [-Wclang-format-violations]
        QVERIFY(FilterParser::parse("fork:true AND archived:false OR name:my-awesome-repo")->evaluate(accessor) == true);
                                                                                                                  ^
tests/TestFilterParser.cpp:145:74: error: code should be clang-formatted [-Wclang-format-violations]
        QVERIFY(FilterParser::parse("john")->evaluate(accessor) == true); // matches owner substring in getAllValues
                                                                         ^
Error: Process completed with exit code 123.
```

File: src/utils/FilterParser.cpp [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+}
+QString InNode::toString() const { return m_key + QStringLiteral(" IN \"") + m_valuesStr + QStringLiteral("\""); }
+
+static bool checkDateFilter(const QString& filterVal, const QString& dateStr, bool isBefore) {
+    QDateTime filterDate = QDateTime::fromString(filterVal, Qt::ISODate);
+    QDateTime dataDate = QDateTime::fromString(dateStr, Qt::ISODate);
+    if (!filterDate.isValid() || !dataDate.isValid()) return false;
+    if (isBefore) return dataDate < filterDate;
+    return dataDate > filterDate;
+}
```
[the remark this answers is not in the document]
Comment (arran4) [reply src/utils/FilterParser.cpp:227]: > medium
> 
> In Qt 6, the overload `QDateTime::fromString(const QString &string, Qt::DateFormat format)` is deprecated. It is recommended to use `QDateTime::fromString` with an explicit format string or use `QDateTime::fromString(string, Qt::ISODateWithMs)` if applicable, or parse it using `QDateTime::fromString(string, Qt::ISODate)` via the non-deprecated overloads if available, or construct it using `QDateTime::fromString(string, QStringLiteral("yyyy-MM-ddTHH:mm:ssZ"))`.

@jules

File: src/RepoListWindow.cpp [hunk 5]
```diff
@@ -54,9 +87,10 @@ void RepoListWindow::setupUI() {
     // Table
     m_table = new QTableWidget(this);
     m_table->setVerticalScrollMode(QAbstractItemView::ScrollPerPixel);
-    m_table->setColumnCount(8);
+    m_table->setColumnCount(11);
     m_table->setHorizontalHeaderLabels({tr("Name"), tr("Owner"), tr("Visibility"), tr("Stars"), tr("Forks"),
-                                        tr("Open Issues"), tr("Updated"), tr("URL")});
+                                        tr("Open Issues"), tr("Created"), tr("Updated"), tr("Archived"), tr("Fork"),
+                                        tr("URL")});
```
[the remark this answers is not in the document]
Comment (arran4) [reply src/RepoListWindow.cpp:93]: > critical
> 
> The column count has been increased to 11, and new columns have been inserted before the URL column. However, the copy action (at line 123) and the custom context menu handler (at line 332) still hardcode column index `7` to retrieve the repository URL. In the new layout, column `7` corresponds to the "Updated" date, and the "URL" column is now at index `10`. This will cause copying or opening the URL to fail or copy the date instead.
> 
> To prevent such bugs and improve maintainability, please define an `enum` for the column indices in `RepoListWindow.h` and use those enum values throughout the class instead of hardcoded magic numbers.

@jules

**Comment by arran4** [comment]
@jules we need to use this view for 
```text
    QList<Variant> variants = {
        {tr("Created by me"), "created_by_me", "author:@me archived:false", "author:@me archived:false",
         "user:@me archived:false"},
        {tr("Assigned to me"), "assigned_to_me", "assignee:@me archived:false", "assignee:@me archived:false", ""},
        {tr("I was mentioned in them"), "mentioned", "mentions:@me archived:false", "mentions:@me archived:false", ""},
        {tr("Review was requested"), "review_requested", "", "review-requested:@me archived:false", ""},
        {tr("Repos I have contributed to"), "contributed", "involves:@me archived:false", "involves:@me archived:false",
         ""},
        {tr("My repos"), "my_repos", "user:@me archived:false", "user:@me archived:false", "user:@me archived:false"},
        {tr("My forks"), "my_forks", "user:@me fork:true archived:false", "user:@me fork:true archived:false",
         "user:@me fork:true archived:false"},
        {tr("Repos I have admin access to"), "admin_access", "user:@me archived:false", "user:@me archived:false",
         "user:@me archived:false"},
        {tr("Archived"), "archived", "archived:true involves:@me", "archived:true involves:@me",
         "archived:true user:@me"},
        {tr("All (Unfiltered)"), "unfiltered", "involves:@me", "involves:@me", "user:@me"}};

```

And allow the filter to be modifiable. Right now when I open them up they don't look like the repo list in the same way. 

Also the RCs are wrong for these custom windows, the xml / rc needs to be included so the menus work .

**Comment by arran4** [comment]
@jules "action" and "tools" menus don't do anything. RC issues? Also can you please ensure that the formula is visible in the other views if it is the thing that is used to filter the input. If not show the api url used.

**Comment by arran4** [comment]
@jules ping

**Comment by arran4** [comment]
> @jules ping
